internal/app/machined/pkg/system/runner/process/process.go

```diff
@@ -105,20 +110,82 @@ func (p *processRunner) build() (cmd *exec.Cmd, logCloser io.Closer, err error)
 		writer = w
 	}
 
-	cmd.Stdout = writer
-	cmd.Stderr = writer
+	var afterStartFuncs []func()
+
+	if p.opts.StdinFile != "" {
+		stdin, err := os.Open(p.opts.StdinFile)
+		if err != nil {
+			return commandWrapper{}, err
+		}
+
+		cmd.Stdin = stdin
+
+		afterStartFuncs = append(afterStartFuncs, func() {
+			stdin.Close() //nolint:errcheck
+		})
+	}
+
+	if p.opts.StdoutFile != "" {
+		stdout, err := os.OpenFile(p.opts.StdoutFile, os.O_WRONLY, 0)
+		if err != nil {
+			return commandWrapper{}, err
+		}
 
-	return cmd, w, nil
+		cmd.Stdout = stdout
+
+		afterStartFuncs = append(afterStartFuncs, func() {
+			stdout.Close() //nolint:errcheck
+		})
+	} else {
+		cmd.Stdout = writer
+	}
+
+	if p.opts.StderrFile != "" {
+		stderr, err := os.OpenFile(p.opts.StderrFile, os.O_WRONLY, 0)
+		if err != nil {
+			return commandWrapper{}, err
+		}
+
+		cmd.Stderr = stderr
+
+		afterStartFuncs = append(afterStartFuncs, func() {
+			stderr.Close() //nolint:errcheck
+		})
+	} else {
+		cmd.Stderr = writer
+	}
+
+	if p.opts.Ctty.IsPresent() {
+		cmd.SysProcAttr = &syscall.SysProcAttr{
+			Setsid:  true,
+			Setctty: true,
+			Ctty:    1,
+		}
+	}
+
+	return commandWrapper{
+		cmd: cmd,
+		afterStart: func() {
+			for _, f := range afterStartFuncs {
+				f()
+			}
+		},
+		afterTermination: func() error {
+			return w.Close()
+		},
+	}, nil
 }
 
 //nolint:gocyclo
 func (p *processRunner) run(eventSink events.Recorder) error {
-	cmd, logCloser, err := p.build()
+	cmdWrapper, err := p.build()
 	if err != nil {
 		return fmt.Errorf("error building command: %w", err)
 	}
 
-	defer logCloser.Close() //nolint:errcheck
+	if cmdWrapper.afterTermination != nil {
```

Comment:
don;t we set the `afterTermination` always here https://github.com/siderolabs/talos/pull/6879/files#diff-439f2aec9d3b4e0f5549166ca67c6cb8735f72c68262955e90185b85f5f41befR173?

Reply:
I was just trying to make it more generic, just in case we set it to be nil later. But I will simplify.

Reply:
done